File: src/main/java/io/playground/orderservice/infrastructure/micrometer/RetryMetricListener.java
package io.playground.orderservice.infrastructure.micrometer;

import io.micrometer.core.instrument.Counter;
import io.micrometer.core.instrument.MeterRegistry;
import lombok.RequiredArgsConstructor;
import org.springframework.context.annotation.Configuration;
import org.springframework.retry.RetryCallback;
import org.springframework.retry.RetryContext;
import org.springframework.retry.RetryListener;

@Configuration
@RequiredArgsConstructor
public class RetryMetricListener implements RetryListener {

    private final MeterRegistry registry;

    @Override
    public <T, E extends Throwable> void onError(RetryContext context, RetryCallback<T, E> callback, Throwable throwable) {
        Counter.builder("retry.attempts")
                .tag("exception",
                        throwable.getClass().getSimpleName())
                .register(registry)
                .increment();
    }
}

File: src/main/kotlin/io/playground/orderservice/infrastructure/micrometer/RetryMetricListener.kt
package io.playground.orderservice.infrastructure.micrometer

import io.micrometer.core.instrument.Counter
import io.micrometer.core.instrument.MeterRegistry
import org.springframework.context.annotation.Configuration
import org.springframework.retry.RetryCallback
import org.springframework.retry.RetryContext
import org.springframework.retry.RetryListener

@Configuration
class RetryMetricListener(
    private val registry: MeterRegistry,
) : RetryListener {
    override fun <T : Any?, E : Throwable?> onError(
        context: RetryContext,
        callback: RetryCallback<T, E>,
        throwable: Throwable,
    ) {
        Counter.builder("retry.attempts")
            .tag("exception", throwable::class.java.simpleName)
            .register(registry)
            .increment()
    }
}
